master/internal/api_config_policies.go

```diff
@@ -35,29 +43,167 @@ func validatePoliciesAndWorkloadType(workloadType, configPolicies string) error
+		constraints = cp.Constraints
+	}
+
+	// Validate constraints if they exist
+	if constraints != nil {
+		if err := a.checkAgainstGlobalPriority(ctx, constraints.PriorityLimit, workloadType); err != nil {
+			return status.Errorf(codes.InvalidArgument, globalPriorityErr)
+		}
+	}
+
+	// Validate against specific configurations
+	if expConfigPolicies != nil {
+		if err := a.validateExperimentConfig(
+			ctx, globalConfigPolicies, expConfigPolicies, constraints, workloadType,
+		); err != nil {
+			return status.Errorf(codes.InvalidArgument, fmt.Sprintf("invalid experiment config policy"+": %s.", err))
+		}
+	}
+
+	if ntscConfigPolicies != nil {
+		if err := a.validateNTSCConfig(ctx, globalConfigPolicies, ntscConfigPolicies, constraints, workloadType); err != nil {
+			return status.Errorf(codes.InvalidArgument, fmt.Sprintf("invalid ntsc config policy"+": %s.", err))
+		}
 	}
 
-	// TODO (request validation): Verify that configs do not violate constraints.
 	return nil
 }
 
-func parseConfigPolicies(configAndConstraints string) (tcps map[string]interface{},
-	invariantConfig *string, constraints *string, err error,
+func (a *apiServer) validateExperimentConfig(
+	ctx context.Context,
+	globalConfigPolicies *model.TaskConfigPolicies,
+	expConfigPolicies *expconf.ExperimentConfigV0,
+	constraints *model.Constraints, workloadType string,
+) error {
+	if expConfigPolicies.RawResources != nil {
+		if err := a.checkAgainstGlobalPriority(ctx, expConfigPolicies.RawResources.RawPriority, workloadType); err != nil {
+			return err
+		}
+		if err := checkConstraintConflicts(constraints, expConfigPolicies.RawResources.RawMaxSlots,
+			expConfigPolicies.RawResources.RawSlotsPerTrial, expConfigPolicies.RawResources.RawPriority); err != nil {
+			return err
+		}
+	}
+
+	return checkAgainstGlobalConfig(ctx, globalConfigPolicies, expConfigPolicies, nil, workloadType)
+}
+
+func (a *apiServer) validateNTSCConfig(
+	ctx context.Context,
+	globalConfigPolicies *model.TaskConfigPolicies,
+	ntscConfigPolicies *model.CommandConfig,
+	constraints *model.Constraints, workloadType string,
+) error {
+	if ntscConfigPolicies.Resources.Priority != nil {
+		if err := a.checkAgainstGlobalPriority(ctx, ntscConfigPolicies.Resources.Priority, workloadType); err != nil {
+			return err
+		}
+	}
+	if err := checkConstraintConflicts(constraints, ntscConfigPolicies.Resources.MaxSlots,
+		&ntscConfigPolicies.Resources.Slots, ntscConfigPolicies.Resources.Priority); err != nil {
+		return err
+	}
+
+	return checkAgainstGlobalConfig(ctx, globalConfigPolicies, nil, ntscConfigPolicies, workloadType)
+}
+
+func (a *apiServer) checkAgainstGlobalPriority(ctx context.Context, taskPriority *int, workloadType string) error {
+	if taskPriority == nil {
+		return nil
+	}
+
+	_, priorityEnabledErr := a.m.rm.SmallerValueIsHigherPriority()
+	if priorityEnabledErr != nil {
+		return fmt.Errorf("task priority is not supported in this cluster: %w", priorityEnabledErr)
```

Comment:
For some of these checks, we want to alert the admin, but still write the config policies to the database. The simplest way is probably to log a warning to master logs. It would be nice if that could also pop up as a toast, but that's probably too ambitious.

Reply:
Sorry for not being clear, this should be a warning, not an error.